### PR TITLE
addHL and subHL affect flags, and are smaller

### PR DESCRIPTION
**Pull request by Clanmaster21**
Most importantly, addHL and subHL now affect the flags as you would expect from a 16 bit addition/subtraction. This seems like it'd be preferred behaviour, however I realise any code relying on it not affecting flags would break. 
One byte saved in addHL, and two bytes saved in subHL. Due to the branching nature of the original code, it's difficult to compare speeds, subHL is either 1 or 6 cycles faster depending on branching, and addHL is between -1 and 3 cycles faster. If we presume a uniform distribution of values in hl, then the chance of branching is a/0xFF, so the expected value of cycles saved is 3*(a/0xFF)-(1-a/0xFF)=4*(a/0xFF)-1. This probably seems like overkill just to say the average of -1 and 3 is 1, but in a critical loop the expected distribution of values could be used to choose which method to inline.

### CODE REVIEW

**Comment by hsoft** [comment]
Thanks again for bringing your optimization skills to the table, it helps a lot.

I'll have to check whether having these routines affect flags (which is an otherwise good idea) is going to cause regressions, which I'll do soon.

**Comment by Clanmaster21** [comment]
No problem, I love this kind of stuff, I'm just happy someone can get some use out of it.
Sounds good, I ran the unit tests (and wrote one to test addHL), so it should at least function.

**Comment by hsoft** [comment]
Something that is good to know regarding tests: To run the tests with a fully updated zasm, you have to run `make updatebootstrap`, otherwise the `zasm/` test subfolder is ran with the old code.

**Comment by Clanmaster21** [comment]
Thank you, I'll keep that in mind for future tests.

**Comment by Clanmaster21** [comment]
Oh wait, I totally messed up my cycle counting, addHL is either 1 or 2 cycles slower depending on branching, and subHL is either 1 or 2 cycles slower too. The bytes saved and the flags stuff still holds though.

**Comment by hsoft** [comment]
This is a tricky one where we need to choose. I'd tend to go with your version, considering that we generally prioritize space over speed.

I've verified that no `addHL` call was in a tight loop and was affected by the flags change. It will make zasm slightly slower (sym register and find inner loop), but not by much.

It seems that `subHL` is unused. I must have removed its only usage at some point.

I had an idea: couldn't we shave off a few cycle in `subHL` with `ld e, a \ xor a \ ld d, a`? `ld` is faster with a register than with a const.

Otherwise, I'd be ready to merge this, but I have one nitpick: please specify that we modify flags according to 16-bit add/sbc. It's clearer that way. Otherwise, one might expect flags from 8-bit operations.

**Comment by hsoft** [comment]
Nevermind my idea earlier, it was a stupid one: we don't want to destroy `A`...

**Comment by Clanmaster21** [comment]
I took a look at the two zasm loops, and I can optimise one of the addHLs and subsequent cpHLDEs, saving 100 cycles and a byte. I can't do much with the other one, but I notice it uses strlen which I can optimise, but I'll put that in another PR. I'll edit the comments too.

**Comment by Clanmaster21** [comment]
Also, it's worth noting that `add hl, de` (and so addHL) only affects the carry and half carry flag, but I'm reluctant to use `adc` as it'd add an extra two bytes and 8 cycles, since we'd need to clear the carry flag too. I also found that `adc hl, de` doesn't set the zero flag in libz80, and someone has opened an issue about it there already, like two years ago.

**Comment by hsoft** [comment]
That's some serious optimisation you're doing all around, thanks! With this and the parsing ones, I can't wait to try whole self-assembly on a RC2014. It used to take 5 minutes.

`(iy-d)` isn't supported? that's an oversight. it should be. I'll add this soon.

Will fully review later today.

**Comment by Clanmaster21** [comment]
No problem, I'm unsure it'll make much difference yet as I have no idea where zasm actually spends most its time, but saving cycles is never a bad thing. 
I don't think so, it seemed to throw an error, and I also remember seeing a comment saying it isn't supported somewhere in the code. I also had trouble doing `ld ixh, a` once, I think, so I don't know if that's supported.

**Comment by hsoft** [comment]
zasm used to support `ixh` and al, but since they are "unofficial" instructions that aren't supported on the Z180, I preferred to drop support for it and not use it anywhere. I should document that.

**Comment by Clanmaster21** [comment]
Ah I see, that explains it, I meant to ask about undocumented instructions but none of them are too useful anyway.

**Comment by Clanmaster21** [comment]
I just ran some numbers, with the RC2014 clocked at around 7MHz, 5 minutes is around 2*10^9 clocks. My new decimal parse saves under 100 clocks for single digit cases, which occurs around 300 times by my count, so that's 30,000 clocks being optimistic, which translates to around 0.004s saved in compilation. There's lots of work to do if we really want to bring down the compilation time, I think that each line compiled is around 10000 clocks, so every 300 clocks per line is a second shaved off compile time.

**Comment by hsoft** [comment]
`parseDecimal` is called more often than there are decimals in the code because zasm, when parsing, starts by parsing for a literal and then falls back to symbols.

**Comment by Clanmaster21** [comment]
I see, I'll have to go back and check to see if time is getting saved all the other times it's called then, I didn't consider that.

File: apps/zasm/symbol.asm [hunk 1]
```diff
@@ -190,9 +192,15 @@ _symFind:
...
-	ld	a, (iy)		; name len again
-	call	addHL		; advance HL by A chars
-	inc	iy \ inc iy \ inc iy
+	
+	push 	de
+	ld 	de, 0x0003
+	add 	iy, de		; faster and shorter than three inc's
+	; offset is signed, but '-' isn't supported by zasm
+	ld 	e, (iy+0xFD)	; offset is also compulsory, so no extra bytes used
+	add 	hl, de	; advance HL by (iy-3) characters
```
Comment (hsoft) [inline apps/zasm/symbol.asm]: You have a wicked mind :) That's a really nice twist. However, I'm afraid that we lost some clarity. Comments don't mention any more that `(iy-3)` refers to name length.

Comment (Clanmaster21) [reply apps/zasm/symbol.asm]: I fixed the comments and also switched it to `(iy-3)` now that it's supported, since that'll also be clearer.

**Comment by hsoft** [comment]
Good stuff, thanks! I'd be ready to merge but I have one last nitpick.

It's hard to find in the docs, but I have this habit of systematically annotate my push/pop that don't wrap whole routines as per https://github.com/hsoft/collapseos/tree/master/kernel#stack-management

If you think it's a bad idea, I'm open to discussion, but otherwise, I'd appreciate if you would follow this convention.

**Comment by Clanmaster21** [comment]
Ah yes, I've seen that, it seems like a good system so I don't mind adopting it.

**Comment by Clanmaster21** [comment]
I think that's right then, I think it's lvl 1 since the only other pushes and pops I see are surrounding the function.

**Pull request merged**